test/alex_koin/slack_commands_test.exs

```diff
@@ -0,0 +1,122 @@
+    end
+  end
+
+  describe "remove_coins" do
+    setup do
+      %{
+        wallet: insert(:wallet)
+      }
+    end
+
+    test "errors when there are no coins", %{wallet: wallet} do
+      assert {:error, :not_enough_coins} = SlackCommands.remove_coins(wallet, 1)
+    end
+
+    test "errors when there are not enough coins", %{wallet: wallet} do
+      insert(:coin, wallet: wallet)
+      assert {:error, :not_enough_coins} = SlackCommands.remove_coins(wallet, 2)
+    end
+
+    test "success if there are enough coins", %{wallet: wallet} do
+      insert(:coin, wallet: wallet)
+      assert :ok = SlackCommands.remove_coins(wallet, 1)
+    end
+  end
+
+  describe "transfer" do
+    setup context do
+      coin_count = Map.get(context, :coin_count, 1)
+      source_wallet = insert(:wallet, balance: coin_count)
+
+      %{
+        source_wallet: source_wallet,
+        target_wallet: insert(:wallet),
+        coins: insert_list(coin_count, :coin, wallet: source_wallet)
+      }
+    end
+
+    def assert_balances([]), do: true
+    def assert_balances([{wallet, amount} | rest]) do
+      float_amount = amount / 1
+
+      assert %Wallet{balance: ^float_amount} = Repo.get(Wallet, wallet.id)
+      assert amount == wallet |> Coin.for_wallet() |> Repo.all |> length
+
+      assert_balances(rest)
+    end
+
+    @tag coin_count: 0
+    test "errors when there are no coins", %{source_wallet: source_wallet , target_wallet: target_wallet} do
+      assert_balances([{source_wallet, 0}, {target_wallet, 0}])
+
+      assert %Wallet{balance: 0.0} = Repo.get(Wallet, source_wallet.id)
+      assert_balances([{source_wallet, 0}, {target_wallet, 0}])
+    end
+
+    @tag coin_count: 1
+    test "fails if wallet has insufficient coins", %{source_wallet: source_wallet , target_wallet: target_wallet} do
+      assert_balances([{source_wallet, 1}, {target_wallet, 0}])
+
+      assert {:error, _, _, _} = SlackCommands.transfer(source_wallet, target_wallet, 2, @memo)
+
+      assert_balances([{source_wallet, 1}, {target_wallet, 0}])
+    end
+
+    @tag coin_count: 1
+    test "transfers coin successfully", %{source_wallet: source_wallet , target_wallet: target_wallet} do
+      assert_balances([{source_wallet, 1}, {target_wallet, 0}])
+
+      assert {:ok, %Transaction{amount: 1.0}} = SlackCommands.transfer(source_wallet, target_wallet, 1, @memo)
+
+      assert_balances([{source_wallet, 0}, {target_wallet, 1}])
+    end
+
+    @tag coin_count: 2
+    test "transfers 2 coins successfully", %{source_wallet: source_wallet , target_wallet: target_wallet} do
+      assert_balances([{source_wallet, 2}, {target_wallet, 0}])
+
+      assert {:ok, %Transaction{amount: 2.0}} = SlackCommands.transfer(source_wallet, target_wallet, 2, @memo)
+
+      assert_balances([{source_wallet, 0}, {target_wallet, 2}])
+    end
+
+    @tag coin_count: 10
+    test "transfers 5 coins successfully", %{source_wallet: source_wallet , target_wallet: target_wallet} do
```

Comment:
The one other thing we should probably verify in a test is the ownership of the coin has changed.

Reply:
It doesn't test the specific coin but does test that some set of coins does actully transfer ownership